src/main/java/io/numaproj/numaflow/sinker/Service.java

```diff
@@ -41,25 +34,58 @@ public Service(Sinker sinker) {
      */
     @Override
     public StreamObserver<SinkOuterClass.SinkRequest> sinkFn(StreamObserver<SinkOuterClass.SinkResponse> responseObserver) {
-        if (this.sinker == null) {
-            return io.grpc.stub.ServerCalls.asyncUnimplementedStreamingCall(
-                    getSinkFnMethod(),
-                    responseObserver);
-        }
+        return new StreamObserver<>() {
+            private boolean startOfStream = true;
+            private CompletableFuture<ResponseList> result;
+            private DatumIteratorImpl datumStream;
+            private boolean handshakeDone = false;
 
-        DatumIteratorImpl datumStream = new DatumIteratorImpl();
+            @Override
+            public void onNext(SinkOuterClass.SinkRequest request) {
+                // make sure the handshake is done before processing the messages
+                if (!handshakeDone) {
+                    if (!request.getHandshake().getSot()) {
+                        responseObserver.onError(new Exception("Handshake request not received"));
+                        return;
+                    }
+                    responseObserver.onNext(SinkOuterClass.SinkResponse.newBuilder()
+                            .setHandshake(request.getHandshake())
+                            .build());
+                    handshakeDone = true;
+                    return;
+                }
 
-        Future<ResponseList> result = sinkTaskExecutor.submit(() -> this.sinker.processMessages(
-                datumStream));
+                // Create a DatumIterator to write the messages to the sinker
+                // and start the sinker if it is the start of the stream
+                if (startOfStream) {
+                    datumStream = new DatumIteratorImpl();
+                    result = CompletableFuture.supplyAsync(
+                            () -> sinker.processMessages(datumStream),
```

Comment:
How are we planning to handle error handling with Future failing?

Reply:
there is a completion exception catch block, which returns error back to the client